pkg/generate/app/cmd/newapp.go

```diff
@@ -1035,15 +1057,59 @@ func (c *AppConfig) checkCircularReferences(objects app.Objects) error {
 				return nil
 			}
 
+			// with reuse of existing image streams, some scenarios have arisen where
+			// comparisons of input/output prior to following the ref to the docker image
+			// are necessary;
+			// ISTs from the same IS with the same tag name, and hence the same docker
+			// image, are still considered circular; however, if the output IST points to the same
+			// image as the image IST because its referential chain takes it to the input IST (i.e. From refs),
+			// where one of the ISTs has a direct docker ref, we are allowing that;
+			// thus, invocations like `oc new-build --binary php`,
+			// where "php" is a image stream with such a structure, do not require
+			// the `--to` option
+			// otherwise, we are deferring to the docker image resolution path
```

Comment:
I don't think we're ever going to be able to catch 100% of circular cases, but i think the main ones we should worry about are:

1) inputIS:Tag==outputIS:Tag
2) inputIS:Tag follows/references outputIS:Tag  (even this is probably not that likely since output is typically :latest and input is not typically :latest because we try to resolve it to something concrete like :7.0)

I think we should probably just stop there.  I'm not even sure we need to do anything else given that new-app always creates (or will now reference an existing) IST for the input and output.  So drilling all the way down to the docker image is unnecessary, i think.

The scenario that would not be covered by the above is:

InputIS:Tag follows MiddleIS:Tag follows outputIS:Tag

Which is certainly not going to be unheard of, but since for the most part the outputIST is going to be following, if anything, not being followed, i'm not that worried about it, but if it's easy enough to check for, we should probably do it.  (this is more than what FollowTagReference currently does, because FollowTagReference does not properly chase tags that reference a different imagestream)

Reply:
Agreed on the never 100% claim.

We certainly cover 1) now.

Yes, the safety valve I introduced does not address 2) directly, but the docker image ref resolution will.

I considered it, but ultimately was really hesitant to torpedo the docker image resolution stuff.  The test case dependencies / scenarios are broad enough that it seems likely customer expectations either steered us down this path, or have manifested as a result.

This effort has certainly put me in a frame of mind where we should take smaller, incremental steps in this space.  

I at least would feel more comfortable soaking this change for a release, but open an item tackle the docker image ref question and multi image stream question (I think you may have to do those in tandem) in the subsequent release.

Reply:
fair enough, i think the dockerimage check is overly aggressive but i haven't (yet) heard of it actually blocking someone who was trying to do something valid.  

and i view the safety valve as largely orthogonal to the "infinite build trigger" scenario.  

essentially the checks are:
1) do you have circular IST refs?  If so we have to error out because your env is broken, and you need to fix those ISTs.  (We need to give a good error when we detect this condition, also)

2) does your output IST match(or otherwise result in updates to) your input IST?  we need to refuse to create this buildconfig because it will trigger itself infinitely once you start it.  Between your new IST1==IST2 check and the dockerimage check this is roughly, but probably still imperfectly, covered.